File: go/hard/minWindow.go
package main

func minWindow(s string, t string) string {
	tm := make(map[byte]int)
	for i := range t {
		tm[t[i]]++
	}

	// 将窗口设置到满足条件的位置
	left := 0
	right := 0
	count := 0
	for right < len(s) {
		if _, e := tm[s[right]]; e {
			if tm[s[right]] > 0 {
				count++
			}
			tm[s[right]]--
		}
		right++
		if count == len(t) {
			break
		}
	}
	if count < len(t) {
		return ""
	}

	// 记录当前满足条件的窗口位置
	start := left
	end := right

	// 开始滑动窗口
	for left < right {
		if v, e := tm[s[left]]; !e { // 如果字符不存在于t中则跳过
			left++
		} else if v < 0 { // 如果字符存在于t中，但当前窗口有多余的该字符，减少该字符数量并跳过
			tm[s[left]]++
			left++
		} else { // 如果当前字符存在于t中，并且当前窗口不存在多余的该字符(即若左边界往右滑动一位会导致不满足条件)
			// 向右寻找字符用于弥补左边界向右滑动一位导致空缺的字符
			for right < len(s) {
				if s[right] == s[left] {
					break
				}
				// 增加窗口中新增的存在于t中的字符
				if _, e := tm[s[right]]; e {
					tm[s[right]]--
				}
				right++
			}
			// 未找到则直接返回
			if right == len(s) {
				break
			}
			right++
			left++
		}

		// 若是当前满足条件的窗口大小小于之前满足条件的窗口大小，则更新
		if right-left < end-start {
			start = left
			end = right
		}
	}
	return s[start:end]
}

func main() {
	minWindow("ADOBECODEBANC", "ABC")
}

File: cpp/hard/minWindow.cpp
#include <functional>
#include <iostream>
#include <queue>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

string minWindow(string s, string t) {
  unordered_map<int, int> tm;
  for (auto c : t) {
    tm[c]++;
  }
  int left = 0;
  while (left < s.size() && tm[s[left]] == 0) {
    left++;
  }
  if (s.size() - left < t.size()) {
    return "";
  }
  int count = 0;
  int right = left;
  queue<int> next;
  while (count < t.size() && right < s.size()) {
    if (tm[s[right]]) {
      next.push(right);
      count++;
      tm[s[right]]--;
      if (count == t.size()) {
        break;
      }
    }
    right++;
  }
  if (count < t.size()) {
    return "";
  }
  string rs = string(s.begin() + left, s.begin() + right + 1);
  while (right < s.size()) {
    right++;
    while (right < s.size() && s[right] != s[left]) {
      right++;
    }
    left = next.front();
    next.pop();
    next.push(right);
    if (right < s.size() && rs.size() > (right + 1 - left)) {
      rs = string(s.begin() + left, s.begin() + right + 1);
    }
  }
  return rs;
}

string minWindow_rain(string s, string t) {
  if (s.size() < t.size())
    return "";
  vector<int> sm(128);
  vector<int> tm(128);
  for (auto c : s) {
    sm[c]++;
  }
  for (auto c : t) {
    tm[c]++;
    if (sm[c] < tm[c])
      return "";
  }
  vector<vector<bool>> visited(s.size(), vector(s.size(), false));
  function<string(int left, int right)> recur = [&](int left,
                                                    int right) -> string {
    if (right + 1 - left < t.size() || visited[left][right]) {
      return "";
    }
    while (tm[s[left]] == 0) {
      left++;
    }
    while (tm[s[right]] == 0) {
      right--;
    }
    string l = "";
    string r = "";

    visited[left][right] = true;
    if (sm[s[left]] > tm[s[left]]) {
      sm[s[left]]--;
      l = recur(left + 1, right);
      sm[s[left]]++;
    }
    if (sm[s[right]] > tm[s[right]]) {
      sm[s[right]]--;
      r = recur(left, right - 1);
      sm[s[right]]++;
    }
    if (l.empty() && r.empty()) {
      return string(s.begin() + left, s.begin() + right + 1);
    }
    if (l.empty()) {
      return r;
    }
    if (r.empty()) {
      return l;
    }
    return l.size() > r.size() ? r : l;
  };
  return recur(0, s.size() - 1);
}

int main() {
  cout << minWindow("ADOBECODEBANC", "ABC");
  return 0;
}
